File: 01-ARRAYS/valid-parentheses.py
from typing import *


# https://leetcode.com/problems/valid-parentheses/
def isValid(s: str) -> bool:
    validBrackets = {'{': '}', '[': ']', '(': ')'}

    openBracketsStack = []
    for c in s:
        if c in validBrackets:
            openBracketsStack.append(c)
            continue
        elif openBracketsStack:
            if c != validBrackets[openBracketsStack[-1]]:
                return False

            openBracketsStack.pop()
            continue

        return False

    return not openBracketsStack

File: 01-ARRAYS/PHP/valid-parentheses.php
<?php

# https://leetcode.com/problems/valid-parentheses/
function isValid(string $s): bool
{
    $validBrackets = [
        '{' => '}',
        '[' => ']',
        '(' => ')',
    ];

    $openBracketsStack = [];
    foreach (str_split($s) as $c) {
        if (array_key_exists($c, $validBrackets)) {
            $openBracketsStack[] = $c;
            continue;
        } elseif (! empty($openBracketsStack)) {
            if ($c != $validBrackets[end($openBracketsStack)])
                return false;
            array_pop($openBracketsStack);
            continue;
        }
        return false;
    }

    return empty($openBracketsStack);
}